### PR TITLE
Stats page

### PR DESCRIPTION
**Pull request by seriouz**
As suggested in bagetter/BaGetter#102 this PR adds a stats page.
I had some trouble with finding out which services are really used during DI setup time. The only thing that came to my mind (without reworking the whole DI registration process) was adding a static helper service for the Statistic that is used to remember which services are acutally in use and not just registered in the DI.

### CODE REVIEW

File: src/BaGetter.Aliyun/AliyunApplicationExtensions.cs [hunk 1]
```diff
@@ -10,8 +12,14 @@ namespace BaGetter;
...
 public static class AliyunApplicationExtensions
 {
-    public static BaGetterApplication AddAliyunOssStorage(this BaGetterApplication app)
+    private const string AliyunOss = "AliyunOss";
+
+    public static BaGetterApplication AddAliyunOssStorage(this BaGetterApplication app, IConfiguration configuration)
     {
+        if (!configuration.HasStorageType(AliyunOss)) return app;
+
+        StatisticsHelperUsedServices.AddServiceToServices(AliyunOss);
```
Comment (loic-sharma) [inline src/BaGetter.Aliyun/AliyunApplicationExtensions.cs:21]: Instead of doing this, could the statistics service load the `StorageOptions` and look at its `Type` property?

File: src/BaGetter.Core/Configuration/StatisticsOptions.cs [hunk 2]
```diff
@@ -0,0 +1,20 @@
+﻿using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+
+namespace BaGetter.Core;
+
+public class StatisticsOptions : IValidatableObject
+{
+    [Required]
+    public bool? HideServices { get; set; } = false;
```
Comment (Regenhardt) [inline src/BaGetter.Core/Configuration/StatisticsOptions.cs:9]: I'd prefer this to just be an additional key `HideStatsPage` in the outer `BaGetterOptions` scope, and if possible optional with a default of false.

File: src/BaGetter.Aliyun/AliyunApplicationExtensions.cs (hunk 1 above)
Comment (Regenhardt) [reply src/BaGetter.Aliyun/AliyunApplicationExtensions.cs:21]: Agreed, just have the service inject `IOptions<BaGetterOptions>` and read the configured types from there. That's less intrusive to the services and configuration calls.

File: src/BaGetter.Core/Configuration/StatisticsOptions.cs (hunk 2 above)
Comment (FroggieFrog) [reply src/BaGetter.Core/Configuration/StatisticsOptions.cs:9]: At the moment I would prefer that too.
I tried to think of other settings for the statistics, but nothing came up. So an extra section seems not justified.

Comment (seriouz) [reply src/BaGetter.Core/Configuration/StatisticsOptions.cs:9]: Sorry I haven't made this clear/obious. This setting (`HideServices`) only hides the listing of used services on the stats page. But after your comments we could add a setting to disable/hide the stats page completely (even from the page menu).

Comment (FroggieFrog) [reply src/BaGetter.Core/Configuration/StatisticsOptions.cs:9]: So if I understand you correct, you want something like this:
```csharp
public class StatisticsOptions : IValidatableObject
{
    [Required]
    public bool? ShowStatisticsPage { get; set; } = true;

    [Required]
    public bool? HideUsedServices { get; set; } = false;
}
```

Comment (seriouz) [reply src/BaGetter.Core/Configuration/StatisticsOptions.cs:9]: Yes, option 1 disables the stats completely and the second option only hide the used services (maybe if you host a public BaGetter).

File: src/BaGetter.Aliyun/AliyunApplicationExtensions.cs (hunk 1 above)
Comment (Regenhardt) [reply src/BaGetter.Aliyun/AliyunApplicationExtensions.cs:21]: The current system allows a configuration change during runtime, without restarting BaGetter, right? I'd like to keep that feature, I think it's nice to be able to change e.g. the search service with zero downtime. Not injecting the service in the first place would break this and force a restart of BaGetter in order to inject the newly configured services.

File: src/BaGetter.Core/Configuration/StatisticsOptions.cs (hunk 2 above)
Comment (Regenhardt) [reply src/BaGetter.Core/Configuration/StatisticsOptions.cs:9]: Ah I see, without services the page still displays version and package stats. Ok that actually makes sense, I'm ok with the extra options then. Maybe name both in the same direction though, either both `Hide` or both `Show`. Or completely different `EnableStatisticsPage` and `IncludeConfiguredServices` (or `DisableStatisticsPage`/`ExcludeConfiguredServices`). I think I like `DisableStatisticsPage`/`ExcludeConfiguredServices` best but it's not a very strong preference. Probably because then the default can be false or null/not included in configuration.

File: docs/docs/configuration.md [hunk 3]
```diff
@@ -203,6 +203,23 @@ This can be useful if you are hosting a private feed and need to host large pack
 }
 ```
 
+## Statistics
```
Comment (FroggieFrog) [inline docs/docs/configuration.md:206]: This section needs to be sdjusted to the latest changes.

File: src/BaGetter.Web/Pages/Statistics.cshtml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+        </div>
+    </div>
+    @if (options.Statistics.ListConfiguredServices is true)
+    {
+        <div class="row">
+            <article class="col-sm-4">
+                <h4>Services in use</h4>
+            </article>
+            <div class="col-sm-8">
+                <h4>@string.Join(", ", IStatisticsService.GetKnownServices())</h4>
```
Comment (FroggieFrog) [inline src/BaGetter.Web/Pages/Statistics.cshtml:51]: Is it beneficial to add more information like "Database: SQLite; Storage: FileSystem"?
Maybe add it in a different PR?

File: docs/docs/configuration.md (hunk 3 above)
Comment (seriouz) [reply docs/docs/configuration.md:206]: Yeah we have to merge the sections, right?

Comment (FroggieFrog) [reply docs/docs/configuration.md:206]: I thought more of adjusting the given example :-)

File: src/BaGetter.Web/Pages/Statistics.cshtml (hunk 4 above)
Comment (seriouz) [reply src/BaGetter.Web/Pages/Statistics.cshtml:51]: This could be a different PR. Also adding the Search provider as service could be done in the PR too.

**Comment by seriouz** [comment]
I don't know what GitHub did here, but it showed merge conflicts but there weren't any during a manual merge.

File: src/BaGetter.Core/Configuration/StatisticsOptions.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
+﻿using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+
+namespace BaGetter.Core;
+
+public class StatisticsOptions : IValidatableObject
+{
+    [Required]
+    public bool? EnableStatisticsPage { get; set; } = false;
```
Comment (Regenhardt) [inline src/BaGetter.Core/Configuration/StatisticsOptions.cs:9]: Why are these required+nullable+default false? Do they have to be nullable if they have a default anyway? Also what does the Required attribute do, is it in case others construct the options themselves?

File: src/BaGetter.Web/Pages/Shared/_Layout.cshtml [hunk 6]
```diff
@@ -27,6 +30,13 @@
                         <ul class="nav navbar-nav" role="tablist">
                             <li role="presentation"><a asp-area="" asp-page="/Index" role="tab" nav-link><span>Packages</span></a></li>
                             <li role="presentation"><a asp-area="" asp-page="/Upload" role="tab" nav-link><span>Upload</span></a></li>
+                            @{
+                                var options = configuration.Get<BaGetterOptions>();
```
Comment (Regenhardt) [inline src/BaGetter.Web/Pages/Shared/_Layout.cshtml:34]: Why not directly inject `IOptionsSnapshot<BaGetterOptions>` instead?

File: src/BaGetter/appsettings.json [hunk 7]
```diff
@@ -53,5 +53,10 @@
 
   "HealthCheck": {
     "Path" : "/health"
+  },
+
+  "Statistics": {
+    "EnableStatisticsPage": false,
```
Comment (Regenhardt) [inline src/BaGetter/appsettings.json:59]: Why is it turned off by default? I'd prefer `true` defaults to show the new page. Services are debatable as that might be an attack vector, but then again BaGetter is mostly used internally so might not be important either.

Comment (seriouz) [reply src/BaGetter/appsettings.json:59]: This makes actual sense. NuGet.org is the same.

File: src/BaGetter.Web/Pages/Shared/_Layout.cshtml (hunk 6 above)
Comment (seriouz) [reply src/BaGetter.Web/Pages/Shared/_Layout.cshtml:34]: done

File: src/BaGetter.Core/Configuration/StatisticsOptions.cs (hunk 5 above)
Comment (seriouz) [reply src/BaGetter.Core/Configuration/StatisticsOptions.cs:9]: Indeed [Required] does nothing at all. And the nullable boolen either.
